File: 6/a.py
import sys
#f = open('test_input')
f = open('input')

coords = []
for l in f:
    parts = list(map(int, l.strip().split(', ')))
    coords.append((parts[0], parts[1]))

minx = min(map(lambda c: c[0], coords))
miny = min(map(lambda c: c[1], coords))
maxx = max(map(lambda c: c[0], coords))
maxy = max(map(lambda c: c[1], coords))

def manhattan(p, c):
    return abs(p[1] - c[1]) + abs(p[0] - c[0])

def onEdge(c):
    (x, y) = c
    return x == minx or x == maxx or y == miny or y == maxy

def getDists(coords):
    dists = {}

    for c in coords:
        #dists[c] = []
        dists[c] = 0

    for y in range(miny, maxy + 1):
        for x in range(minx, maxx + 1):
            p = (x, y)
            minManhattan = sys.maxsize
            minC = ()
            equal = False
            for c in coords:
                dist = manhattan(p, c)
                if dist == minManhattan:
                    equal = True
                if dist < minManhattan:
                    minManhattan = dist
                    minC = c
                    equal = False
            if not equal:
                dists[minC] += 1
            if onEdge(p):
                dists[minC] = -sys.maxsize
    return dists

def solve(coords):
    dists = getDists(coords)
    maxL = 0
    for d in dists:
        length = dists[d]
        if length > maxL:
            maxL = length
    return maxL

print(solve(coords))
        

File: 6/a.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>

typedef struct Point {
    uint32_t x;
    uint32_t y;
} Point;

typedef struct Corners {
    Point min;
    Point max;
} Corners;

uint32_t manhattan(Point* p1, Point* p2) {
    return abs(p1->x - p2->x) + abs(p1->y - p2->y);
}

bool isInfinite(Point *c, Corners* corners) {
    return c->x == corners->min.x 
    || c->x == corners->max.x      
    || c->y == corners->min.y
    || c->y == corners->max.y;
}

int main() {
    FILE* f = fopen("input", "r");
    //FILE* f = fopen("test_input", "r");
    int num_coords = 0;
    while (!feof(f)) {
        char ch = fgetc(f);
        if (ch == '\n') num_coords++;
    }
    Point* coords = calloc(num_coords, sizeof(Point));
    uint32_t* closest = calloc(num_coords, sizeof(uint32_t));
    bool* infinite = calloc(num_coords, sizeof(bool));
    rewind(f);
    uint32_t x, y;
    Corners corners = (Corners) {{UINT32_MAX, UINT32_MAX}, {0, 0}};

    for (int i = 0 ; i < num_coords ; i++) {
        fscanf(f, "%d, %d", &x, &y);
        coords[i] = (Point){x, y};
        if (x < corners.min.x) corners.min.x = x;
        if (y < corners.min.y) corners.min.y = y;
        if (x > corners.max.x) corners.max.x = x;
        if (y > corners.max.y) corners.max.y = y;
    }
    
    for (int y = corners.min.y ; y <= corners.max.y ; y++) {
        for (int x = corners.min.x ; x <= corners.max.x ; x++) {
            Point p = (Point){x, y};
            uint32_t min_manhattan = UINT32_MAX;
            int closest_index;
            bool equal = false;
            for (int i = 0 ; i < num_coords ; i++) {
                uint32_t dist = manhattan(&p, &(coords[i]));
                //printf("dist: %d\n", dist);
                if (dist == min_manhattan) {
                    equal = true;
                } else if (dist < min_manhattan) {
                    min_manhattan = dist;
                    closest_index = i;
                    equal = false;
                }
            }
            if (infinite[closest_index]) continue;
            if (!equal) {
                closest[closest_index]++;
            } 
            if (isInfinite(&p, &corners)) {
                infinite[closest_index] = true;
            }
        }
    }
    uint32_t max_length = 0;
    for (int i = 0 ; i < num_coords ; i++) {
        if (closest[i] > max_length && !infinite[i]) {
            max_length = closest[i];
        }
    }
    printf("%d\n", max_length);
    return 0;
}
